libplatsupport/src/mach/eswin/serial.c

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2023, UNSW
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <stdlib.h>
+#include <string.h>
+#include <platsupport/serial.h>
+#include <platsupport/plat/serial.h>
+#include "../../chardev.h"
+
+#define UART_THR 0x00 /* UART Transmit Holding Register */
+#define UART_IER 0x04 /* UART Interrupt Enable Register */
+#define UART_IER_ERDAI BIT(0) /* Enable Received Data Available Interrupt */
+#define UART_LSR 0x14 /* UART Line Status Register */
+#define UART_LSR_THRE 0x20 /* Transmit Holding Register Empty */
+
+#define REG_PTR(base, off)     ((volatile uint32_t *)((base) + (off)))
+
+int uart_getchar(ps_chardevice_t *d)
+{
+    while ((*REG_PTR(d->vaddr, UART_LSR) & BIT(0)));
+    return *REG_PTR(d->vaddr, UART_THR);
+}
+
+int uart_putchar(ps_chardevice_t *d, int c)
+{
+    if (c == '\n' && (d->flags & SERIAL_AUTO_CR)) {
+        uart_putchar(d, '\r');
+    }
+
+    while ((*REG_PTR(d->vaddr, UART_LSR) & UART_LSR_THRE) == 0);
```

Comment:
Probably better to check whether the buffer is not full, if there is a queue larger than one byte in the device.

(Ideally printf implementations would wait till the buffer is empty before writing a new line and keep transmitting characters while the buffer has space once they started. That way you can share a UART with multiple users like a Linux VM or other cores pretty well, without it becoming a mess. But oh well.)

Reply:
Oh well, never mind then.

Reply:
Reading the linked pdf, you can poll `TEMT` to see whether any data is queued and poll `THRE` before queueing consequent bytes. That would avoid the interleaved output if everyone would do this.

Anyway, with the above scheme `THRE` is used as buffer not empty already, so the code as it is is fine.